src/hotspot/share/gc/shared/oopStorage.cpp

```diff
@@ -880,67 +880,52 @@ bool OopStorage::should_report_num_dead() const {
-//
-// We don't want cleanup work to linger indefinitely, but we also don't want
-// to run the service thread too often.  We're also very limited in what we
-// can do in a release operation, where cleanup work is created.
-//
+
 // When a release operation changes a block's state to empty, it records the
 // need for cleanup in both the associated storage object and in the global
-// request state.  A safepoint cleanup task notifies the service thread when
+// request state. The ServiceThread checks at timed intervals if
 // there may be cleanup work for any storage object, based on the global
-// request state.  But that notification is deferred if the service thread
-// has run recently, and we also avoid duplicate notifications.  The service
-// thread updates the timestamp and resets the state flags on every iteration.
+// request state.  We don't want to run empty block cleanup too often in the
+// face of frequent explicit ServiceThread wakeups, hence the defer period.
 
 // Global cleanup request state.
 static volatile bool needs_cleanup_requested = false;
 
-// Flag for avoiding duplicate notifications.
-static bool needs_cleanup_triggered = false;
+// Time after which a cleanup is permitted.
+static jlong cleanup_permit_time = 0;
 
-// Time after which a notification can be made.
-static jlong cleanup_trigger_permit_time = 0;
-
-// Minimum time since last service thread check before notification is
-// permitted.  The value of 500ms was an arbitrary choice; frequent, but not
-// too frequent.
-const jlong cleanup_trigger_defer_period = 500 * NANOSECS_PER_MILLISEC;
-
-void OopStorage::trigger_cleanup_if_needed() {
-  MonitorLocker ml(Service_lock, Monitor::_no_safepoint_check_flag);
-  if (Atomic::load(&needs_cleanup_requested) &&
-      !needs_cleanup_triggered &&
-      (os::javaTimeNanos() > cleanup_trigger_permit_time)) {
-    needs_cleanup_triggered = true;
-    ml.notify_all();
-  }
-}
+// Minimum time between ServiceThread cleanups.
+// The value of 500ms was an arbitrary choice; frequent, but not too frequent.
+const jlong cleanup_defer_period = 500 * NANOSECS_PER_MILLISEC;
 
 bool OopStorage::has_cleanup_work_and_reset() {
   assert_lock_strong(Service_lock);
-  cleanup_trigger_permit_time =
-    os::javaTimeNanos() + cleanup_trigger_defer_period;
-  needs_cleanup_triggered = false;
-  // Set the request flag false and return its old value.
-  // Needs to be atomic to avoid dropping a concurrent request.
-  // Can't use Atomic::xchg, which may not support bool.
-  return Atomic::cmpxchg(&needs_cleanup_requested, true, false);
+
+  if (Atomic::load_acquire(&needs_cleanup_requested) &&
+      os::javaTimeNanos() > cleanup_permit_time) {
+    cleanup_permit_time =
+      os::javaTimeNanos() + cleanup_defer_period;
+    // Set the request flag false and return its old value.
+    // Needs to be atomic to avoid dropping a concurrent request.
+    Atomic::release_store(&needs_cleanup_requested, false);
```

Comment:
The comment above seems to imply that needs_cleanup_requested can be set to true atomically and that we therefore have to use Atomic::xchg to make sure we don't drop a concurrent request for cleanup. Now the code has been changed to a release_store instead. If the comment is still right (I think it is?), then the code shouldchange back to xchg. Otherwise the comment should be updated to describe why we don't need xchg any longer, I think.

Reply:
That comment is a left-over from the old mechanism, and is no longer true.  This is the only place
that sets it false (and is holding Service_lock so there can't be concurrent setters to false), and knows
it is true (from the earlier test).  The old protocol with the safepoint cleanup trigger was more complex,
and this function didn't have the test of the flag.  We probably don't even need a "release" here.

Reply:
I thought the comment about "avoid dropping a concurrent request" was rather talking about the flag being concurrently set from false to true (in record_needs_cleanup), but then immediately getting cleared to false here due to the lack of atomics, leading to the cleanup request being essentially ignored and handled as done, even though nothing was done.

Reply:
If the flag was false at the preceding load-acquire then there wasn't a notification (yet) so we don't
attempt to do any work this time around.  But there's always next (now at least periodic) time.  It was
more complicated before the ServiceThread became periodic.  (Possibly overly complicated.  While
discussing these changes with Coleen offline I had trouble understanding some of the old interaction
and thought there was a much simpler way to accomplish what it was trying to do.  But making the
ServiceThread periodic allows even more simplification.)

Reply:
Here, if needs_cleanup_requested was true and then we set it to false, we still return true so a concurrent thread setting it to true is ok because we still return 'true'.  That is, we need to do some work.   If we return false here, while another thread is setting it to true, we'll get it on the next periodic timeout.
The only reason for the store-release is because we use load-acquire and it's consistent. The comment refers to the old behaviour which was more complicated.  I removed the comment to remove the confusion.

Reply:
Ok, I think I restated what Kim said.

Reply:
> The only reason for the store-release is because we use load-acquire and it's consistent.

So why do we use load-acquire? That is typically to pair with a store-release. We only need release semantics if someone seeing this store must see previous stores as well.

Reply:
Ah, okay. Then this seems fine. Maybe tweak or remove the comment to reflect this conversation. It seems to be saying the opposite of what it is doing.

Reply:
I removed the comment that used to describe the CAS.